File: Others/NOS/DP/5.笛卡兒積DP/P2098.py
"""
P2098 [USACO16DEC] Team Building P
https://www.luogu.com.cn/problem/P2098
"""
MOD = int(1e9 + 9)
N, M, K = map(int, input().split())
A = list(map(int, input().split()))
B = list(map(int, input().split()))
A.sort()
B.sort()

def solve1():  # Memoization (MLE)
    from functools import cache
    import sys
    sys.setrecursionlimit(int(5e3))
    @cache
    def f(i: int, j: int, k: int) -> int:
        if k == 0:
            return 1
        if i < 0 or j < 0:
            return 0
        res = (f(i - 1, j, k) + f(i, j - 1, k) - f(i - 1, j - 1, k)) % MOD
        if A[i] > B[j]:
            res += f(i - 1, j - 1, k - 1)
            res %= MOD
        return res
    print(f(N - 1, M - 1, K))

def solve2a():  # 查表法 (MLE)
    f = [[[0] * (K + 1) for _ in range(M + 1)] for _ in range(N + 1)]
    for i in range(N + 1):
        for j in range(M + 1):
            f[i][j][0] = 1
    for i in range(1, N + 1):
        for j in range(1, M + 1):
            for k in range(1, K + 1):
                f[i][j][k] = (f[i - 1][j][k] + f[i][j - 1][k] - f[i - 1][j - 1][k]) % MOD
                if A[i - 1] > B[j - 1]:
                    f[i][j][k] += f[i - 1][j - 1][k - 1]
                    f[i][j][k] %= MOD
    print(f[N][M][K])

def solve2b():  # 查表法 (Rolling Array)
    f = [[[1] + [0] * K for _ in range(M + 1)] for _ in range(2)]
    for i in range(1, N + 1):
        cur, pre = i & 1, (i - 1) & 1
        for j in range(1, M + 1):
            for k in range(1, K + 1):
                f[cur][j][k] = (f[pre][j][k] + f[cur][j - 1][k] - f[pre][j - 1][k]) % MOD
                if A[i - 1] > B[j - 1]:
                    f[cur][j][k] += f[pre][j - 1][k - 1]
                    f[cur][j][k] %= MOD
    print(f[N & 1][M][K])

# solve1()
# solve2a()
solve2b()

File: Others/NOS/DP/5.笛卡兒積DP/P2098.cpp
/*
 * P2098 [USACO16DEC] Team Building P
 * https://www.luogu.com.cn/problem/P2098
 */
#include <bits/stdc++.h>
using namespace std;
using i64 = int64_t;
const int MOD = 1e9 + 9;
#define endl '\n'

void solve1() {
    int N, M, K;
    cin >> N >> M >> K;
    vector<int> A(N), B(M);
    for (int i = 0; i < N; i++) cin >> A[i];
    for (int i = 0; i < M; i++) cin >> B[i];
    sort(A.begin(), A.end());
    sort(B.begin(), B.end());

    vector<vector<vector<i64>>> memo(N + 1, vector<vector<i64>>(M + 1, vector<i64>(K + 1, -1)));
    auto dfs = [&](this auto&& dfs, int i, int j, int k) -> i64 {
        if (k == 0) return 1;
        if (i < 0 || j < 0) return 0;
        i64& res = memo[i][j][k];
        if (res != -1) return res;
        res = (dfs(i - 1, j, k) + dfs(i, j - 1, k) - dfs(i - 1, j - 1, k) + MOD) % MOD;
        if (A[i] > B[j]) res = (res + dfs(i - 1, j - 1, k - 1)) % MOD;
        return res;
    };
    cout << dfs(N - 1, M - 1, K) << endl;
    return;
}

void solve2() {
    int N, M, K;
    cin >> N >> M >> K;
    vector<int> A(N), B(M);
    for (int i = 0; i < N; i++) cin >> A[i];
    for (int i = 0; i < M; i++) cin >> B[i];
    sort(A.begin(), A.end());
    sort(B.begin(), B.end());

    vector<vector<vector<i64>>> f(N + 1, vector<vector<i64>>(M + 1, vector<i64>(K + 1, 0)));
    for (int i = 0; i <= N; i++)
        for (int j = 0; j <= M; j++)
            f[i][j][0] = 1;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            for (int k = 1; k <= K; k++) {
                f[i][j][k] = (f[i - 1][j][k] + f[i][j - 1][k] - f[i - 1][j - 1][k] + MOD) % MOD;
                if (A[i - 1] > B[j - 1]) f[i][j][k] = (f[i][j][k] + f[i - 1][j - 1][k - 1]) % MOD;
            }
        }
    }
    cout << f[N][M][K] << endl;
    return;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    solve1();
    // solve2();
    return 0;
}